example/snake/qmodel.cpp

```diff
@@ -0,0 +1,79 @@
+//
+// lager - library for functional interactive c++ programs
+// Copyright (C) 2019 Carl Bussey
+//
+// This file is part of lager.
+//
+// lager is free software: you can redistribute it and/or modify
+// it under the terms of the MIT License, as detailed in the LICENSE
+// file located at the root of this source code distribution,
+// or here: <https://github.com/arximboldi/lager/blob/master/LICENSE>
+//
+
+#include "actions.hpp"
+#include "qmodel.hpp"
+
+namespace sn {
+
+namespace {
+
+QPoint as_qpoint(std::pair<int, int> p)
+{
+    auto [x, y] = p;
+    return {x, y};
+}
+
+} // namespace
+
+QSnakeBody::QSnakeBody(QObject* parent)
+    : QAbstractListModel(parent)
+{}
+
+void QSnakeBody::update(snake_t::body_t body)
+{
+    if (data_ != body) {
+        beginResetModel();
+        data_ = body;
+        endResetModel();
+    }
+}
+
+int QSnakeBody::rowCount(const QModelIndex&) const { return data_.size(); }
+
+QVariant QSnakeBody::data(const QModelIndex& index, int) const
+{
+    if (index.row() < data_.size())
+        return as_qpoint(data_[index.row()]);
+    return {};
+}
+
+QGame::QGame(context_t context)
+    : context_{std::move(context)}
+    , snakeBody_{new QSnakeBody(this)}
+{}
+
+void QGame::update(game_t game)
```

Comment:
If we call the other function `update()` this could be called `view`. You could consider having all these qt based models get a reader cursor and do the watching themselves. Would simplify some of this code for sure.

Reply:
I've renamed the other `update`. `view` felt a little bit strange to me as a member function. From the object oriented perspective, it feels a little bit confusing, as it doesn't describe how I'm interacting with the object. (From a functional perspective, it makes sense.)

As mentioned in a previous comment - I'll definitely have a play around more with the cursors, but I wanted to keep them out of this example.